envoy/http/codec.h

```diff
@@ -174,6 +175,19 @@ class ResponseEncoder : public virtual StreamEncoder {
    * @param decoder new request decoder.
    */
   virtual void setRequestDecoder(RequestDecoder& decoder) PURE;
+
+  /**
+   * Set headers, trailers, and stream info for deferred logging.
+   * @param request_header_map Request headers for this stream.
+   * @param response_header_map Response headers for this stream.
+   * @param response_trailer_map Response trailers for this stream.
+   * @param stream_info Stream info for this stream.
+   */
+  virtual void
+  setDeferredLoggingHeadersAndTrailers(Http::RequestHeaderMapSharedPtr request_header_map,
+                                       Http::ResponseHeaderMapSharedPtr response_header_map,
+                                       Http::ResponseTrailerMapSharedPtr response_trailer_map,
+                                       StreamInfo::StreamInfo& stream_info) PURE;
```

Comment:
You probably already discussed this, but FWIW, I find this dance a bit odd. It's a bit confusing how the headers are passed back and forth now. Perhaps more comments? Also, if this is for logging, can we use a shared_ptr to const object here to make it more clear that these things are not modified after this point?

I do wonder if it would have been easier to understand to keep the deferred logging system inside the HCM (some kind of deferred logging manager) somehow triggered by a codec callback later, but that would be a bigger change so probably not worth it.

Reply:
> Also, if this is for logging, can we use a shared_ptr to const object here to make it more clear that these things are not modified after this point?

Done.

> Perhaps more comments?

I'll take a pass at more thorough comments/docstrings. Is there anything specifically/particularly confusing that I should clear up?

Reply:
I've added a few more comments on the methods that pass headers through to the codec.